crates/ruff/src/rules/refurb/rules/repeated_append.rs

```diff
@@ -0,0 +1,377 @@
+/// `append` re-sizes the list individually, whereas `extend` re-sizes it once for all elements.
+///
+/// ## Example
+/// ```python
+/// nums = [1, 2, 3]
+///
+/// nums.append(4)
+/// nums.append(5)
+/// nums.append(6)
+/// ```
+///
+/// Use instead:
+/// ```python
+/// nums = [1, 2, 3]
+///
+/// nums.extend((4, 5, 6))
+/// ```
+///
+/// ## References
+/// - [Python documentation: More on Lists](https://docs.python.org/3/tutorial/datastructures.html#more-on-lists)
+#[violation]
+pub struct RepeatedAppend {
+    name: String,
+    replacement: String,
+}
+
+impl Violation for RepeatedAppend {
+    const AUTOFIX: AutofixKind = AutofixKind::Sometimes;
+
+    #[derive_message_formats]
+    fn message(&self) -> String {
+        let RepeatedAppend { name, replacement } = self;
+        format!("Use `{replacement}` instead of repeatedly calling `{name}.append()`")
+    }
+
+    fn autofix_title(&self) -> Option<String> {
+        let RepeatedAppend { replacement, .. } = self;
+        Some(format!("Replace with `{replacement}`"))
+    }
+}
+
+pub(crate) fn convert_consecutive_appends_into_extend(checker: &mut Checker, stmt: &Stmt) {
+    let Some(appends) = match_consecutive_appends(checker, stmt) else {
+        return;
+    };
+
+    // No need to proceed if we have less than 1 append to work with.
+    if appends.len() <= 1 {
+        return;
+    }
+
+    let groups = group_appends(&appends);
+
+    // group borrows from checker, so we can't directly push into checker.diagnostics
+    let mut diagnostics: Vec<Diagnostic> = vec![];
+    diagnostics.reserve_exact(groups.len());
+
+    for group in groups {
+        let replacement = make_suggestion(&group, checker.generator());
+        let start = group.appends.first().unwrap().stmt.start();
+        let end = group.appends.last().unwrap().stmt.end();
+
+        let mut diagnostic = Diagnostic::new(
+            RepeatedAppend {
+                name: group.name().to_string(),
+                replacement: replacement.to_string(),
+            },
+            TextRange::new(start, end),
+        );
+
+        // We only suggest fix when all appends in a group are clumped together.
+        if checker.patch(diagnostic.kind.rule()) && group.is_consecutive {
+            diagnostic.set_fix(Fix::suggested(Edit::replacement(replacement, start, end)));
+        }
+
+        diagnostics.push(diagnostic);
+    }
+    checker.diagnostics.extend(diagnostics);
+}
+
+#[derive(Debug, Clone)]
+struct Append<'a> {
+    receiver: &'a ExprName,
+    binding: &'a Binding<'a>,
+    argument: &'a Expr,
+    stmt: &'a Stmt,
+}
+
+#[derive(Debug)]
+struct AppendGroup<'a> {
+    appends: Vec<Append<'a>>,
+    is_consecutive: bool,
+}
+
+impl<'a> AppendGroup<'a> {
+    fn name(&self) -> &'a str {
+        &self.appends.first().unwrap().receiver.id
+    }
+}
+
+fn match_consecutive_appends<'a>(checker: &'a Checker, stmt: &'a Stmt) -> Option<Vec<Append<'a>>> {
```

Comment:
Nit: can we use an empty `Vec` rather than the `Option` state?

Reply:
Sure!